### PR TITLE
Implements PE checksum and signing validation

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

**Comment by tmat** [comment]
Supersedes https://github.com/dotnet/corefx/pull/11458

**Comment by tmat** [comment]
FYI: @agocke @jaredpar 
Once this change is in I'll update SRM Roslyn depends on to the latest, so that we can use the new APIs.

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs [hunk 1]
```diff
@@ -409,5 +411,162 @@ private static int SumRawDataSizes(ImmutableArray<SerializedSection> sections,Se
...
+            writer.WriteBytes(signature);
+        }
+
+        // internal for testing
+        internal static uint CalculateChecksum(BlobBuilder peImage, Blob checksumFixup)
+        {
+            return CalculateChecksum(GetContentToChecksum(peImage, checksumFixup)) + (uint)peImage.Count;
+        }
+
+        private unsafe static uint CalculateChecksum(IEnumerable<Blob> blobs)
```
Comment (agocke) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: The checksum is only 32 bits? Is this CRC?

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs [hunk 2]
```diff
@@ -409,5 +411,162 @@ private static int SumRawDataSizes(ImmutableArray<SerializedSection> sections,Se
...
+        {
+            uint checksum = 0;
+            int pendingByte = -1;
+
+            foreach (var blob in blobs)
+            {
+                var segment = blob.GetBytes();
+                fixed (byte* arrayPtr = segment.Array)
+                {
+                    Debug.Assert(segment.Count > 0);
```
Comment (agocke) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: Move out of `fixed`?

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs (hunk 1 above)
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: AFAICT, the PE/COFF spec just says "The image file checksum. The algorithm for computing the checksum is incorporated into IMAGHELP.DLL". Should we push to have this algorithm documented? Should we add the documentation to our Addendum to that spec in this repo in the meantime?

**Comment by nguerrera** [comment]
:+1:

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs (hunk 1 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: Yeah, let's follow up with adding it to Addendum.

Comment (agocke) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: Also, maybe consider a struct wrapper to represent the checksum type.

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: I don't think a struct is necessary - this is private API used in one place.

File: src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs (hunk 2 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs]: I'll do that in a separate change.

**Pull request merged**